ramls/instance-preceding-succeeding-titles.raml

```diff
@@ -67,6 +67,40 @@ resourceTypes:
           body:
             text/plain:
               example: "Internal server error, contact administrator"
+    put:
+      description: Update preceding/succeeding titles related to the instance
+      queryParameters:
+        instanceId:
+          type: string
+          required: true
+          pattern: "^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[1-5][a-fA-F0-9]{3}-[89abAB][a-fA-F0-9]{3}-[a-fA-F0-9]{12}$"
```

Comment:
Why was a query parameter chosen instead of including the instance in the path?

Reply:
That's because in the case of adding it to `instance-preceding-succeeding-titles` interface it will violate the REST principle to go from general to more specific in the url.

Reply:
> That's because in the case of adding it to `instance-preceding-succeeding-titles` interface it will violate the REST principle to go from general to more specific in the url.

I don't think I'm aware of that principal, where can I learn more about it?

How does using a query parameter avoid this challenge?

Reply:
https://restful-api-design.readthedocs.io/en/latest/urls.html
Titles are more specific then instance. So instance should come first and titles after. But it's not possible in current interface. That's why instance id is in query params

Reply:
> https://restful-api-design.readthedocs.io/en/latest/urls.html

Thanks for the link, I wasn't aware of this guidance at all.

I had a quick read of that document and couldn't find anything about going from general to specific ideas in a URL. Which section of the document talks about this that I missed?

> Titles are more specific then instance. So instance should come first and titles after. 

What do you mean by `more specific`? Do you mean that a preceding / succeeding title is beneath an instance in a hierarchy?

Does that mean that `/instances/{id}/preceding-succeeding-titles` would be ok however `/preceding-succeeding-titles/instances/{id}/` is not?

And so `preceding-succeeding-titles?instanceId={id}` was chosen instead to try to remove the hierarchy from the path?

> But it's not possible in current interface. That's why instance id is in query params

Why is it not possible? Is it because Okapi stops the use of instance in the path in this interface because it's used in another interface?

Reply:
> I had a quick read of that document and couldn't find anything about going from general to specific ideas in a URL. Which section of the document talks about this that I missed?

There is about collection and sub-collection hierarchy:
![image](https://user-images.githubusercontent.com/5904625/128003978-5979507b-0815-42a2-a953-9b5f46776389.png)

> What do you mean by more specific? Do you mean that a preceding / succeeding title is beneath an instance in a hierarchy?

Yes, you understood me right. Preceding / succeeding title couldn't exist without instance so it is beneath an instance in a hierarchy.

> Does that mean that /instances/{id}/preceding-succeeding-titles would be ok however /preceding-succeeding-titles/instances/{id}/ is not?

Yes.

> Why is it not possible? Is it because Okapi stops the use of instance in the path in this interface because it's used in another interface?

It's not possible because we can't use `/preceding-succeeding-titles/instances/{id}` because of collection/sub-collection principle. And we can't use smth like `/instances/{id}/preceding-succeeding-titles` in this interface because all paths in the interface should be consistent (in our case starts with `/preceding-succeeding-titles`

Reply:
Thank you for sharing your reasoning with me.

> > I had a quick read of that document and couldn't find anything about going from general to specific ideas in a URL. Which section of the document talks about this that I missed?
> 
> There is about collection and sub-collection hierarchy:
> ![image](https://user-images.githubusercontent.com/5904625/128003978-5979507b-0815-42a2-a953-9b5f46776389.png)

Thank you for referring to specifically to that. I think I interpreted that less strongly than you did.

> > What do you mean by more specific? Do you mean that a preceding / succeeding title is beneath an instance in a hierarchy?
> 
> Yes, you understood me right. Preceding / succeeding title couldn't exist without instance so it is beneath an instance in a hierarchy.

Ok, 

> > Does that mean that /instances/{id}/preceding-succeeding-titles would be ok however /preceding-succeeding-titles/instances/{id}/ is not?
> 
> Yes.
> 
> > Why is it not possible? Is it because Okapi stops the use of instance in the path in this interface because it's used in another interface?
> 
> It's not possible because we can't use `/preceding-succeeding-titles/instances/{id}` because of collection/sub-collection principle. 

Like the above, I think I didn't interpret that guidance that sub-collections _must_ be nested under a non-collection resource.

bq. And we can't use smth like `/instances/{id}/preceding-succeeding-titles` in this interface because all paths in the interface should be consistent (in our case starts with `/preceding-succeeding-titles`

Why do all of the paths need to be consistent and start with the same path?

Reply:
The pros of REST approach is that developer could expect uniform interface from API. If paths are not consistent then it becomes chaos. 

So, why are you against the approach to have instance id as a query parameter? Could you provide some cons of it?

Reply:
> The pros of REST approach is that developer could expect uniform interface from API. If paths are not consistent then it becomes chaos.

My understanding is the uniform interface principle refers more to how a client interacts with resources, rather than the structure of specific URLs.

> So, why are you against the approach to have instance id as a query parameter? Could you provide some cons of it?

I don't think I've argued against the use of a query parameter. I wanted to understand the reasoning behind the decision.

I am biased for paths. I find paths easier to comprehend and easier to construct than query parameters. And that query parameters can easily be missed or ignored.

I found the idea of a [clean URL](https://en.wikipedia.org/wiki/Clean_URL) as a good explanation of my bias (that admittedly, I developed prior to knowing about that concept).

Reply:
This query parameter is required so it couldn't be missed or ignored.

But I heard you. You are the owner of this module, so I will agree with any your decision. What URL do you prefer to have for this endpoint?

Reply:
> This query parameter is required so it couldn't be missed or ignored.

Fair enough.

> But I heard you. You are the owner of this module, so I will agree with any your decision. What URL do you prefer to have for this endpoint?

I'm not trying to make an argument from authority (though it seems I have). I wanted to understand the reasoning. 

Neither am I'm not attempting to block the (already approved) work or request changes. I don't think there is a need to change the code again at this point. I will resolve this comment.

As for what I would do, there are some challenges with how the preceding / succeeding titles are already structured. Given that, my preference would be for a path something like `preceding-succeeding-titles/instances/{id}`.
If there had been a prefix for this interface, I might have gone with `preceding-succeeding-titles-storage/instances/{id}/preceding-succeeding-title`.

Each of these have challenges with long paths and are somewhat clunky. If FOLIO didn't have the business logic split and we were starting from scratch, I'd probably use something like `instances/{id}/preceding-titles` and `instances/{id}/succeeding-titles` to represent these two collections (though that is somewhat misleading as these resources aren't really independent, due to how preceding / succeeding titles can be associated with more than one instance.

Reply:
It's not a big deal to change it. I have done this.